File: LeetCode/Search-Insert-Position/C++/SearchInsertPosition.cpp
#include <vector>
#include <iostream>
using namespace std;

class Solution
{
public:
    int searchInsert(vector<int> &nums, int target) 
    {
        int start = 0; // initialize start index of binary search
        int end = nums.size() - 1; // initialize end index of binary search

        while (start <= end) {
             // perform binary search
            int mid = start + (end - start) / 2; // calculate mid index

            if (nums[mid] == target) { 
                // if element found at mid, return mid
                return mid;
            }
            else if(nums[mid] < target){
                // if element at mid is less than target, then target would be inserted after mid
                start = mid + 1;
            }
            else {
                // if element at mid is greater than target, then target would be inserted before mid
                end = mid - 1;
            }
    }
      // If target is not found and start is greater thsn end, it means target would be inserted at start index
    return start; 
    }
};

int main() {
    Solution solution; // Create object of Solution class
    vector<int> nums = {1,3,5,6}; // Example input array
    int target = 5; // Target element to search
    int result = solution.searchInsert(nums, target);
    cout << "The index of the target element is: " << result << endl; // Outputs the index of the target element or the index where it needs to be inserted
return 0;
}

File: LeetCode/Search-Insert-Position/Python/search_insert_position.py
class Solution(object):
    def searchInsert(self, nums, target):
        start, end = 0, len(nums) - 1

        while start <= end:
            mid = start + (end - start) // 2

            if nums[mid] == target:
                return mid  # Target found at index mid
            elif nums[mid] < target:
                start = mid + 1  # Target is in the end half
            else:
                end = mid - 1  # Target is in the start half

        return start  # Target not found, return the insertion point
# Create an instance of the Solution class
solution = Solution()

# Example usage:
nums = [1, 3, 5, 6]
target = 5
result = solution.searchInsert(nums, target)
print("Result:", result)  # Output: 2
